akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+
+  /**
+   * Marker trait for remote messages serialized by
+   * [[akka.cluster.ddata.protobuf.ReplicatorMessageSerializer]].
+   */
+  trait ReplicatorMessage extends Serializable
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] object Internal {
+
+    case object GossipTick
+    case object RemovedNodePruningTick
+    case object ClockTick
+    final case class Write(key: String, envelope: DataEnvelope) extends ReplicatorMessage
+    case object WriteAck extends ReplicatorMessage
+    final case class Read(key: String) extends ReplicatorMessage
+    final case class ReadResult(envelope: Option[DataEnvelope]) extends ReplicatorMessage
+    final case class ReadRepair(key: String, envelope: DataEnvelope)
+    case object ReadRepairAck
+    final case class BufferedCommand(cmd: Command, replyTo: ActorRef)
+    final case class UpdateInProgress(cmd: Update[ReplicatedData], replyTo: ActorRef)
+
+    // Gossip Status message contains SHA-1 digests of the data to determine when
+    // to send the full data
+    type Digest = ByteString
+    val DeletedDigest: Digest = ByteString.empty
+    val LazyDigest: Digest = ByteString(0)
+    val NotFoundDigest: Digest = ByteString(-1)
+
+    final case class DataEnvelope(
+      data: ReplicatedData,
+      pruning: Map[UniqueAddress, PruningState] = Map.empty)
+      extends ReplicatorMessage {
+
+      import PruningState._
+
+      def needPruningFrom(removedNode: UniqueAddress): Boolean =
+        data match {
+          case r: RemovedNodePruning ⇒ r.needPruningFrom(removedNode)
+          case _                     ⇒ false
+        }
+
+      def initRemovedNodePruning(removed: UniqueAddress, owner: UniqueAddress): DataEnvelope = {
+        copy(pruning = pruning.updated(removed, PruningState(owner, PruningInitialized(Set.empty))))
+      }
+
+      def prune(from: UniqueAddress): DataEnvelope = {
+        data match {
+          case dataWithRemovedNodePruning: RemovedNodePruning ⇒
+            require(pruning.contains(from))
+            val to = pruning(from).owner
+            val prunedData = dataWithRemovedNodePruning.prune(from, to)
+            copy(data = prunedData, pruning = pruning.updated(from, PruningState(to, PruningPerformed)))
+          case _ ⇒ this
+        }
+
+      }
+
+      def merge(other: DataEnvelope): DataEnvelope =
+        if (other.data == DeletedData) DeletedEnvelope
+        else {
+          var mergedRemovedNodePruning = other.pruning
+          for ((key, thisValue) ← pruning) {
+            mergedRemovedNodePruning.get(key) match {
+              case None ⇒ mergedRemovedNodePruning = mergedRemovedNodePruning.updated(key, thisValue)
+              case Some(thatValue) ⇒
+                mergedRemovedNodePruning = mergedRemovedNodePruning.updated(key, thisValue merge thatValue)
+            }
+          }
+
+          copy(data = cleaned(data, mergedRemovedNodePruning), pruning = mergedRemovedNodePruning).merge(other.data)
```

Comment:
perhaps add a comment that confirms that the intent is to clean up both sides before merging, it took me a while to completely understand this—or inline the merging here